File: Agent Chain - Pipeline Pattern/langgraph_python_example/agent_chain.py
# Agent Chain / Pipeline Pattern - LangGraph Python Example

This example demonstrates the Agent Chain / Pipeline Pattern using LangGraph and Python. Three agents (Extractor, Summarizer, Translator) process input sequentially. The LLM is Mistral.

## Requirements

- Python 3.8+
- `langgraph` library
- `python-dotenv` (for .env support)
- Mistral LLM API access

## Install dependencies

```bash
pip install langgraph python-dotenv requests
```

## Example Code

```python
import os
from langgraph import Agent, Environment, LLM
from dotenv import load_dotenv

load_dotenv()

MISTRAL_API_KEY = os.getenv("MISTRAL_API_KEY")
MISTRAL_API_URL = "https://api.mistral.ai/v1/chat/completions"

class SimpleEnvironment(Environment):
    def get_observation(self):
        return input("Input: ")
    def send_action(self, action):
        print(action)

class Extractor(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, input_text):
        prompt = f"You are Alice, an Extractor. Extract the key information from: {input_text}"
        return self.llm.complete(prompt)

class Summarizer(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, extracted):
        prompt = f"You are Bob, a Summarizer. Summarize this: {extracted}"
        return self.llm.complete(prompt)

class Translator(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, summary):
        prompt = f"You are Carol, a Translator. Translate this to French: {summary}"
        return self.llm.complete(prompt)

llm = LLM(
    provider="mistral",
    api_key=MISTRAL_API_KEY,
    api_url=MISTRAL_API_URL,
)

env = SimpleEnvironment()
extractor = Extractor(llm)
summarizer = Summarizer(llm)
translator = Translator(llm)

input_text = env.get_observation()
extracted = extractor.act(input_text)
print("Alice (Extractor):", extracted)
summarized = summarizer.act(extracted)
print("Bob (Summarizer):", summarized)
translated = translator.act(summarized)
print("Carol (Translator):", translated)
```

---

- Try a paragraph or document to see the pipeline in action.
- Make sure your `.env` file contains your Mistral API key.

File: Agent Chain - Pipeline Pattern/autogen_typescript_example/agent_chain.ts
// Agent Chain / Pipeline Pattern - Autogen TypeScript Example
// To run: npm install && npm run agent-chain

import axios from 'axios';
import * as readline from 'readline';
import * as dotenv from 'dotenv';
dotenv.config();

const MISTRAL_API_URL = 'https://api.mistral.ai/v1/chat/completions';
const MISTRAL_API_KEY = process.env.MISTRAL_API_KEY;

async function agent(name: string, role: string, input: string): Promise<string> {
  const prompt = `You are ${name}, your role is: ${role}. Here is your input: ${input}`;
  const response = await axios.post(
    MISTRAL_API_URL,
    {
      model: 'mistral-tiny',
      messages: [{ role: 'user', content: prompt }],
    },
    {
      headers: {
        'Authorization': `Bearer ${MISTRAL_API_KEY}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

async function agentChain(task: string) {
  // Agent 1: Extractor
  const extracted = await agent('Alice', 'Extractor', task);
  console.log('Alice (Extractor):', extracted);

  // Agent 2: Summarizer
  const summarized = await agent('Bob', 'Summarizer', extracted);
  console.log('Bob (Summarizer):', summarized);

  // Agent 3: Translator
  const translated = await agent('Carol', 'Translator (to French)', summarized);
  console.log('Carol (Translator):', translated);
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('Input: ', async (userInput: string) => {
  try {
    await agentChain(userInput);
  } catch (err) {
    console.error('Error:', err);
  }
  rl.close();
});
